add "name" and "addr" props to the FDT, to more easily act like other platforms

richlowe opened the pull request:
There's an assumption in common code that prom device tree nodes have a "name" property.  In the FDT nodes are named inherently, not by a property, and so we have to fake it.

Rather than trying to fake it in getprop/setprop/etc as was done originally just add pseudo properties to the tree as we build it.

We also add the "addr" property, which is used to name child devices in the actual device tree, but could later be replaced with something else.

This fixes `prtconf -pd`, the problem with which was that prom_nextprop did _not_ insert fake entries for "name" and "addr", unlike get/set, and so the verbose openprom snapshot did not contain names, though the non-verbose did, and without names all `prtconf` can do is print "no data available" once per node.

I have not tested that this works with PCI as we don't have any platform with pci except raspberrypi 4 (which I don't have), and even then we don't enumerate PCI, so unless it does, there are no nodes to lookup anyway.  I haven't fixed the endianness issue that probably exists, because I can't test it.

I'd be grateful if @citrus-it or @hadfl could test on rpi4, and could take a glance, and additionally if @rmustacc could take a glance.

Further, @rmustacc asked in the past about the name of our root nexus.  I happened to notice that at present we don't name the root prom tree node.  i86pc calls this i86pc, I don't remember what SPARC uses.  Opinions on what we should do?  For now, the empty string seems fine.

(I haven't fixed nits in this yet, in case someone has a problem with the method.  I will before I merge)

richlowe committed:
Add "name" and "unit-address" props to the FDT, to more closely match 1275

According to the devicetree specification[1] what in 1275 would be the "name"
and "unit-address" properties of a prom node are instead reflected in the
inherent name of the node (which 1275 does not have).

Rather than trying to fake these properties in `prom_getprop`, `prom_setprop`,
etc. just add these properties to the tree as we build it.

[1]: https://devicetree-specification.readthedocs.io/en/v0.3/devicetree-basics.html#properties

rmustacc commented:
Right now on our systems the root prom node and the root devinfo node are the same. I suspect this is more of a side effect of how we "emulate" the prom on the oxide architecture. For example, here are the first few entries:

```
unknown # prtconf -p
System Configuration:  Oxide Computer Company  oxide
Memory size: 131055 Megabytes
System Peripherals (PROM Nodes):

Node 'Oxide,Gimlet'
    Node 'ramdisk'
    Node 'huashan'
    Node 'pci'
        Node 'pci1022,1480'
        Node 'pci1022,164f'
        Node 'pci1022,1482'
```

From the work I did in [illumos#15529 Improve libtopo product detection](https://www.illumos.org/issues/15529), we do use the devinfo (not prom) name as a way to further disambiguate things when there is no SMBIOS present. I expect that this will be prominent for some ARM systems, though also insufficient given the way that many of them are ultimately used.

I had a few different prtconf -vp outputs from SPARC systems from folks when I was looking at some of this. Here, they do have names that come from the product. Here are a few examples from different systems:

```
Node 0xf0229e94
    idprom:  01840014.4f46ec9a.00000000.46ec9ade.00000000.00000000.00000000.00000000
    scsi-initiator-id:  00000007
    banner-name:  'T5140'
    name:  'SUNW,T5140'
    stick-frequency:  457656f0
    hv-api-groups:  00000000.00000001.00000100.00000002.00000003.00000101.00000108
    breakpoint-trap:  0000007f 
    device_type:  'sun4v'
    compatible: 'sun4v'
    #address-cells:  00000002
    #size-cells:  00000002
```

```
Node 0xf0229e94
    idprom:  01860010.e0769180.00000000.76918077.00000000.00000000.00000000.00000000
    scsi-initiator-id:  00000007
    pcie-acs-supported:  
    pcie-ari-supported:
    banner-name:  'SPARC T4-1'
    name:  'ORCL,SPARC-T4-1'
    stick-frequency:  3b8f5850 
    hv-api-groups:  00000000.00000001.00000100.00000002.00000204.00000003.00000101.00000105.00000108.00000107
    breakpoint-trap:  0000007f
    device_type:  'sun4v'
    compatible: 'sun4v'
    #address-cells:  00000002
    #size-cells:  00000002
```

This is an ldom:

```
Node 0xf0229e94
    idprom:  01840014.4ffbe576.00000000.fbe576de.00000000.00000000.00000000.00000000
    scsi-initiator-id:  00000007
    pcie-acs-supported:  
    pcie-ari-supported:  
    banner-name:  'SPARC T4-1'
    name:  'ORCL,SPARC-T4-1'
    stick-frequency:  3b8f5850
    hv-api-groups:  00000000.00000001.00000100.00000002.00000204.00000003.00000101.00000105.00000108.00000107
    breakpoint-trap:  0000007f
    device_type:  'sun4v'
    compatible: 'sun4v'
    #address-cells:  00000002
    #size-cells:  00000002
```

I think we can start with the empty string, but it seems like trying to match a product name longer term would be advantageous, though given the way that these different boards will work, I expect we'll still need disambiguation from the name to figure out either particular implementations (that is to say to distinguish two RPI4s) since they'll have wildly different GPIOs, iomux configs, etc. that we'll want to use and may even be user specific.

richlowe committed:
name the root prom node after the implementation

This matches other platforms, most especially SPARC but also x86 (though the
implementation names there do not vary).

richlowe commented:
I force pushed this (I assume that's ok, with no source-level comments), to both name "unit-address" correctly, to make sure that nothing outside of the prom implementation can change the "name" or "unit-address" properties, and to add a name property to the / node with the value matching the implementation architecture (as happens on SPARC by way of 1275, and is copied on x86)

richlowe commented:
Breaks raspberry pi, fixed as per #34 instead
